### ISSUE TITLE
connect() sometimes fails to fire error event when failing to connect

### ISSUE THREAD

**Issue by teo1978** [peers/peerjs#329]
Attempt to connect to a peer that doesn't exist, and listen for the `error` event.
Expected: should fire the `error` event.
Observed: sometimes it does, sometimes NOTHING. It fails to connect (which is expected), but it does not fire the error event either. No `open`, `close` or `error` event, nothing.

Example code:

```
    var peer;
    var conn;
    var otherId="whatever";
    var connected=false;

    $(function(){
      peer=new Peer({ 
        host: 'example.com',
        port: 9000,
        key: 'xxx'
      });

      startConnect();
    });

    function startConnect() {
        connected=false;
        $("#connection-info").html("Connecting...");
        conn=peer.connect(otherId);
        conn.on('open', function(){
            console.log("connection open");
            initDataConnection();
        })
        conn.on('close', function(){
            console.log("Connection closed");
            connected=false;
            startConnect();
        });
        conn.on('error', function() {
            connected=false;
            $("#connection-info").html("Connection error.");
            console.log("connection error ");
        });
    }

    function initDataConnection() {
      connected=true;
      conn.on('data', function(data){
        console.log("Received: "+data);
      });

    }

```

**Comment by ndac-todoroki** [comment]
I think your code is not right.
The listen which catches errors of `peer.connect(otherId)` is 

```javascript 
peer.on('error', function(err) { ... });
 ```

but not `DataConnection.on('error', function(err) { ... });`, which you are implementing as

```javascript
conn = peer.connect(otherId);  // conn is a DataConnection object
conn.on('error', function() {
    connected=false;
    $("#connection-info").html("Connection error.");
    console.log("connection error ");
});
```

But I don't know what `DataConnection.on('error', function(err) { ... });` listens to exactly

**Comment by teo1978** [comment]
> The listen which catches errors of peer.connect(otherId) is `peer.on('error'...)`

Oh, I see. My bad.

However, `DataConnection` sometimes *does* trigger `error` events because my code catches them. I guess that's when an error occurs on a connection that is already established (as opposed to an error when trying to connect).

But this is only speculation, because
> But I don't know what DataConnection.on('error', function(err) { ... }); listens to exactly

Nor do I, because it's not documented.

**Comment by ndac-todoroki** [comment]
If your error triggers in the Example above, and if you're only handling connections "to a peer that doesn't exist", I think it is a weird thing. Can you show some specific cases?

**Comment by teo1978** [comment]
No, when I connect to a peer that does exist and then the peer closes the connection (or disappears or whatever), the error triggers. When I connect to a peer that doesn't exist, it doesn't.

Since you pointed out that I should listen on the Peer and not on the DataConnection for the error on establishing the connection, all is pretty much as expected.

My original report was bogus because my code was flawed; the only remaining issue here is the fact that the documentation says nothing about the `error` event of `DataConnection` (except for the fact that it exists)

**Comment by teo1978** [comment]
No, WAIT A MOMENT.

**I have modified my code**. Now it definitely **should trigger errors** when failing to connect for whatever reason, and it not always does.

Updated code:
```
    var peer;
    var conn;
    var otherId="whatever";
    var connected=false;
    
    $(function(){
      peer=new Peer({ 
        host: 'example.com',
        port: 9000,
        key: 'xxx'
      });
      
      // START ADDED CODE <<<<<<<<<<<<<<<
      peer.on('error', function() {
      	connected=false;
	$("#connection-info").html("Connection error.");
	console.log("connection error "+connErrors);
      });
      // END ADDED CODE <<<<<<<<<<<<<<<<<
      
      startConnect();
    });
    
    function startConnect() {
    	connected=false;
		$("#connection-info").html("Connecting...");
		conn=peer.connect(otherId);
		conn.on('open', function(){
			console.log("connection open");
			initDataConnection();
		})
		conn.on('close', function(){
			console.log("Connection closed");
			connected=false;
			startConnect();
		});
		conn.on('error', function() {
			connected=false;
			$("#connection-info").html("Connection error.");
			console.log("connection error ");
		});
    }
    
    function initDataConnection() {
      connected=true;
      conn.on('data', function(data){
        console.log("Received: "+data);
      });
      
    }
    
```

Expected: with the above code, it should **never**, ever happen that in the console you see "connecting..." and then nothing more. After "connecting...", either you should see "connection open" if the connection succeeds, or "connection error" if it fails.

Observed: that does happen. Right now, I am using the above code to connect to a peer (which, by the way, is supposed to be listening, so I don't even know why the connection is not succeeding in the first place, but i might have an error on the other peer's code failing to report that it has stopped listening - but that's not relevant here), and I get "Connecting..." and then nothing. And no JS errors.

**Comment by ndac-todoroki** [comment]
@teo1978, if you run the code and observed the same next time, can you paste what `peer` and `conn` returns in the console?
Also, it might be saying something if you create your peer with `debug:3` option. I'll try this too...since it'll be very serious on my app too.

**Comment by teo1978** [comment]
Any news? This bug is very serious.

What I have been able to figure out so far is that, if I try to connect to a non-existing peer, neither the Peer object (which is the one which should) neither the DataConnection object fire any error event.

Here is the output from the Peer attempting to connect to a non-existent remote peer, with debug:3

`
peer.min.js:1 PeerJS:  Creating RTCPeerConnection.
peer.min.js:1 PeerJS:  Listening for ICE candidates.
peer.min.js:1 PeerJS:  Listening for `negotiationneeded`
peer.min.js:1 PeerJS:  Listening for data channel
peer.min.js:1 PeerJS:  Listening for remote stream
peer.min.js:1 PeerJS:  `negotiationneeded` triggered
peer.min.js:1 PeerJS:  Created offer.
peer.min.js:1 PeerJS:  Set localDescription: offer for: qaq
2peer.min.js:1 PeerJS:  Received ICE candidates for: qaq
peer.min.js:1 PeerJS:  Socket open
`

**Comment by teo1978** [comment]
The only workaround is to start a timer when trying to connect. If you don't get the `open` event after a certain time, assume the connection has failed. Which is quite pathetic.

**Comment by teo1978** [comment]
Oh, I don't know why, but I have repeated the **exact same test** without changing anything, and now there is an uncaught error (as opposed to nothing happening at all):

```
peer.min.js:1 PeerJS:  Creating RTCPeerConnection.
peer.min.js:1 PeerJS:  Listening for ICE candidates.
peer.min.js:1 PeerJS:  Listening for `negotiationneeded`
peer.min.js:1 PeerJS:  Listening for data channel
peer.min.js:1 PeerJS:  Listening for remote stream
peer.min.js:1 PeerJS:  `negotiationneeded` triggered
peer.min.js:1 PeerJS:  Created offer.
peer.min.js:1 PeerJS:  Set localDescription: offer for: qaq
2peer.min.js:1 PeerJS:  Received ICE candidates for: qaq
?peerid=qaq:172 Peer created
peer.min.js:1 PeerJS:  Socket open
peer.min.js:1 PeerJS:  ERROR Error: Could not connect to peer qaq
?peerid=qaq:175 Uncaught ReferenceError: err is not defined(anonymous function) @ ?peerid=qaq:175d.emit @ peer.min.js:1c.emitError @ peer.min.js:1c._handleMessage @ peer.min.js:1(anonymous function) @ peer.min.js:1d.emit @ peer.min.js:1_socket._socket.onmessage @ peer.min.js:1
```

**Comment by teo1978** [comment]
Sorry for the previous deleted comments. I had some errors in my test code.

Are there any news? I still confirm the issue and I can't find a precise pattern. This bug is **VERY SERIOUS**, as not being able to reliably detect when a connection cannot be established or is lost renders peerjs totally unusable.

Here's a recap:
- I have a Peer object that tries to connect to a remote peer via `connect()`
- I listen to the `error` event on **both** the Peer object itself and the DataConnection object returned by `connect`. I also listen to the `close` event on the DataConnection object.
- **usually, as expected**, if I try to connect to a remote Peer that can't be reached (e.g. it doesn't exist), an `error` event is fired by the Peer object
- **usually, as expected**, if the remote Peer goes away, a `close` event is fired either by the DataConnection object.
- However, **at random times**, I call `connect()`, and **neither the `open` nor the `error` event are fired**.

The way I systematically reproduce the issue is as follow:
- Whenever I detect an `error` event either in the Peer or the DataConnection object, or a `close` event on the DataConnection, I set a timeout and start all over, creating a new Peer and trying to connect it to the remote, whose id is known.
- I start with the remote peer listening. So the connection is succesful at the first try. The connection is established, and the `open` event is fired as expected
- Then I take the remote down. So, my Peer detects the `close` event as expected and starts retrying.
- it keeps detecting the `error` event as expected and keeps retrying
- then I restart the remote which start listening again

At this point, it would be expected that the connection from the local Peer succeeds. Instead, it neither succeeds neither triggers an error. Neither the `open` event (on DataConnection) nor the `error` event (on the Peer) are fired.

**Comment by arsenalol69** [comment]
I had the same issue ! Any solution please ? As said before, it's a very serious problem !

**Comment by pantrombka** [comment]
I have this problem when I use node-webrtc with peerjs. I can't find the reason.

**Comment by skyne98** [comment]
Any solutions?

**Comment by ndac-todoroki** [comment]
maybe using nttcom's fork(https://github.com/skyway/skyway-js-sdk), or doing some patch may be a solution: This repo seems not maintained.

**Comment by afrokick** [comment]
Should be fixed in 1.0.4+
